addons/website_sale_stock/models/product_template.py

```diff
@@ -24,31 +52,53 @@ def _get_combination_info(self, combination=False, product_id=False, add_qty=1,
+            incoming_qty = product_with_context.incoming_qty
+            incoming_date = product_with_context._get_next_incoming_move_date()
+            uom_name = product.uom_id.name
+            cart_qty = product.cart_qty
+            remaining_qty_formatted = self.env['ir.qweb.field.float'].value_to_html(free_qty - cart_qty, {'precision': 0})
+
             combination_info.update({
-                'virtual_available': virtual_available,
-                'virtual_available_formatted': self.env['ir.qweb.field.float'].value_to_html(virtual_available, {'precision': 0}),
+                'free_qty': free_qty,
                 'product_type': product.type,
-                'inventory_availability': product.inventory_availability,
-                'available_threshold': product.available_threshold,
-                'custom_message': product.custom_message,
-                'product_template': product.product_tmpl_id.id,
-                'cart_qty': product.cart_qty,
-                'uom_name': product.uom_id.name,
+                'allow_to_order': self.allow_to_order,
+                'incoming_qty': incoming_qty,
+                'incoming_date': fields.Date.to_string(incoming_date),
+                'available_threshold': self.available_threshold,
+                'product_template': self.id,
+                'cart_qty': cart_qty,
+                'uom_name': uom_name,
+                'show_availability': self.show_availability,
+                'available_message': self._get_stock_message('available').format(qty=remaining_qty_formatted, unit=uom_name),
+                'out_of_stock_message': self._get_stock_message('out_of_stock').format(qty=remaining_qty_formatted, unit=uom_name),
+                'threshold_message': self._get_stock_message('threshold').format(qty=remaining_qty_formatted, unit=uom_name),
             })
         else:
             product_template = self.sudo()
             combination_info.update({
-                'virtual_available': 0,
+                'free_qty': 0,
                 'product_type': product_template.type,
-                'inventory_availability': product_template.inventory_availability,
+                'allow_to_order': product_template.allow_to_order,
                 'available_threshold': product_template.available_threshold,
-                'custom_message': product_template.custom_message,
                 'product_template': product_template.id,
-                'cart_qty': 0
+                'cart_qty': 0,
             })
 
         return combination_info
+
+    def _is_sold_out(self):
+        combination_info = self.with_context(website_sale_stock_get_quantity=True)._get_combination_info()
+        return combination_info['product_type'] == 'product' and combination_info['free_qty'] <= 0
+
+    def _get_website_ribbon(self):
+        if self._is_sold_out():
+            return self.env['product.ribbon'].search([('id', '=', self.env.ref('website_sale.sold_out_ribbon').id)])
```

Comment:
Isn't `self.env.ref('website_sale.sold_out_ribbon')` by itself already the same thing ?

Reply:
Yes it is... Thanks !